## Telecom Churn: Decision Trees with PCA

With 226 predictor variables, we need to predict whether a particular customer will switch to another telecom provider or not. In telecom terminology, customer attrition is referred to as 'churn'

In [122]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [123]:
churn_data = pd.read_csv("telecom_churn_data.csv")

### Reading Data from the DataSet

In [124]:
churn_data.info()
print('\n')
print(churn_data.columns,'\n')
print(churn_data.describe(), '\n')
churn_data.head(10)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99999 entries, 0 to 99998
Columns: 226 entries, mobile_number to sep_vbc_3g
dtypes: float64(179), int64(35), object(12)
memory usage: 172.4+ MB


Index(['mobile_number', 'circle_id', 'loc_og_t2o_mou', 'std_og_t2o_mou',
       'loc_ic_t2o_mou', 'last_date_of_month_6', 'last_date_of_month_7',
       'last_date_of_month_8', 'last_date_of_month_9', 'arpu_6',
       ...
       'sachet_3g_9', 'fb_user_6', 'fb_user_7', 'fb_user_8', 'fb_user_9',
       'aon', 'aug_vbc_3g', 'jul_vbc_3g', 'jun_vbc_3g', 'sep_vbc_3g'],
      dtype='object', length=226) 

       mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
count   9.999900e+04    99999.0         98981.0         98981.0   
mean    7.001207e+09      109.0             0.0             0.0   
std     6.956694e+05        0.0             0.0             0.0   
min     7.000000e+09      109.0             0.0             0.0   
25%     7.000606e+09      109.0             0.0             0.0   

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
0     7000842753        109             0.0             0.0             0.0   
1     7001865778        109             0.0             0.0             0.0   
2     7001625959        109             0.0             0.0             0.0   
3     7001204172        109             0.0             0.0             0.0   
4     7000142493        109             0.0             0.0             0.0   
5     7000286308        109             0.0             0.0             0.0   
6     7001051193        109             0.0             0.0             0.0   
7     7000701601        109             0.0             0.0             0.0   
8     7001524846        109             0.0             0.0             0.0   
9     7001864400        109             0.0             0.0             0.0   

  last_date_of_month_6 last_date_of_month_7 last_date_of_month_8  \
0            6/30/2014            7/31/2014            8/31/2014   
1            6/30/2014            7/31/2014            8/31/2014   
2            6/30/2014            7/31/2014            8/31/2014   
3            6/30/2014            7/31/2014            8/31/2014   
4            6/30/2014            7/31/2014            8/31/2014   
5            6/30/2014            7/31/2014            8/31/2014   
6            6/30/2014            7/31/2014            8/31/2014   
7            6/30/2014            7/31/2014            8/31/2014   
8            6/30/2014            7/31/2014            8/31/2014   
9            6/30/2014            7/31/2014            8/31/2014   

  last_date_of_month_9    arpu_6     ...      sachet_3g_9  fb_user_6  \
0            9/30/2014   197.385     ...                0        1.0   
1            9/30/2014    34.047     ...                0        NaN   
2            9/30/2014   167.690     ...                0        NaN   
3            9/30/2014   221.338     ...                0        NaN   
4            9/30/2014   261.636     ...                0        0.0   
5            9/30/2014    50.258     ...                0        NaN   
6            9/30/2014   429.023     ...                0        NaN   
7            9/30/2014  1069.180     ...                0        NaN   
8            9/30/2014   378.721     ...                0        NaN   
9            9/30/2014   119.518     ...                0        NaN   

   fb_user_7  fb_user_8  fb_user_9   aon  aug_vbc_3g  jul_vbc_3g  jun_vbc_3g  \
0        1.0        1.0        NaN   968       30.40        0.00      101.20   
1        1.0        1.0        NaN  1006        0.00        0.00        0.00   
2        NaN        NaN        1.0  1103        0.00        0.00        4.17   
3        NaN        NaN        NaN  2491        0.00        0.00        0.00   
4        NaN        NaN        NaN  1526        0.00        0.00        0.00   
5        NaN        NaN        NaN  1471        0.00        0.00        0.00   
6        NaN        NaN        NaN  1673        0.00        0.00        0.00   
7        NaN        NaN        NaN   802       57.74       19.38       18.74   
8        1.0        1.0        NaN   315       21.03      910.65      122.16   
9        1.0        NaN        NaN   902        0.00        0.00        0.00   

   sep_vbc_3g  
0        3.58  
1        0.00  
2        0.00  
3        0.00  
4        0.00  
5        0.00  
6        0.00  
7        0.00  
8        0.00  
9        0.00  

[10 rows x 226 columns]

### Cleaning the Data 
#### Datatype Conversion  
#### Missing Value Treatment
#### Scaling the Data

### Creating Derived Variables

### Filtering High Value Customers based on 70 Percentile Recharge Criteria in September

In [125]:
churn_data['AverageRecharge67'] = ((churn_data['total_rech_amt_6']+churn_data['total_rech_amt_7'])/2)
churn_data['High Value'] = np.where(churn_data.total_rech_amt_9>=churn_data.AverageRecharge67.quantile(0.7),1,0)
highvalue=churn_data[churn_data['High Value'] == 1]
highvalue.info()
print('\n')
print(highvalue.describe(),'\n')
highvalue.head(10)



<class 'pandas.core.frame.DataFrame'>
Int64Index: 29010 entries, 3 to 99997
Columns: 228 entries, mobile_number to High Value
dtypes: float64(180), int32(1), int64(35), object(12)
memory usage: 50.6+ MB


       mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
count   2.901000e+04    29010.0         28939.0         28939.0   
mean    7.001228e+09      109.0             0.0             0.0   
std     6.840121e+05        0.0             0.0             0.0   
min     7.000000e+09      109.0             0.0             0.0   
25%     7.000651e+09      109.0             0.0             0.0   
50%     7.001245e+09      109.0             0.0             0.0   
75%     7.001815e+09      109.0             0.0             0.0   
max     7.002411e+09      109.0             0.0             0.0   

       loc_ic_t2o_mou        arpu_6        arpu_7        arpu_8        arpu_9  \
count         28939.0  29010.000000  29010.000000  29010.000000  29010.000000   
mean              0.0    475.

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
3      7001204172        109             0.0             0.0             0.0   
15     7002070365        109             0.0             0.0             0.0   
21     7002124215        109             0.0             0.0             0.0   
24     7001125315        109             0.0             0.0             0.0   
33     7000149764        109             0.0             0.0             0.0   
38     7000815202        109             0.0             0.0             0.0   
48     7000294396        109             0.0             0.0             0.0   
53     7002188521        109             0.0             0.0             0.0   
54     7001957064        109             0.0             0.0             0.0   
57     7001682585        109             0.0             0.0             0.0   

   last_date_of_month_6 last_date_of_month_7 last_date_of_month_8  \
3             6/30/2014            7/31/2014            8/31/2014   
15            6/30/2014            7/31/2014            8/31/2014   
21            6/30/2014            7/31/2014            8/31/2014   
24            6/30/2014            7/31/2014            8/31/2014   
33            6/30/2014            7/31/2014            8/31/2014   
38            6/30/2014            7/31/2014            8/31/2014   
48            6/30/2014            7/31/2014            8/31/2014   
53            6/30/2014            7/31/2014            8/31/2014   
54            6/30/2014            7/31/2014            8/31/2014   
57            6/30/2014            7/31/2014            8/31/2014   

   last_date_of_month_9    arpu_6     ...      fb_user_7  fb_user_8  \
3             9/30/2014   221.338     ...            NaN        NaN   
15            9/30/2014    31.000     ...            NaN        NaN   
21            9/30/2014   514.453     ...            NaN        NaN   
24            9/30/2014   422.050     ...            NaN        NaN   
33            9/30/2014   977.020     ...            1.0        1.0   
38            9/30/2014   363.987     ...            NaN        NaN   
48            9/30/2014  1873.271     ...            NaN        NaN   
53            9/30/2014   978.077     ...            0.0        0.0   
54            9/30/2014    82.866     ...            NaN        NaN   
57            9/30/2014  1201.298     ...            1.0        1.0   

    fb_user_9   aon  aug_vbc_3g  jul_vbc_3g  jun_vbc_3g  sep_vbc_3g  \
3         NaN  2491        0.00        0.00        0.00         0.0   
15        NaN   270        0.00        0.00        0.00         0.0   
21        NaN   720        0.00        0.00        0.00         0.0   
24        NaN  3072        0.00        0.00        0.00         0.0   
33        1.0  1328      358.10     2706.83     1714.67         0.0   
38        0.0  1155        0.00        0.00        0.00         0.0   
48        NaN  1265        0.00        0.00        0.00         0.0   
53        0.0   697        0.00        0.00        0.00         0.0   
54        NaN   538        0.00        0.00        0.00         0.0   
57        1.0  3106     1799.09     3649.39     1892.32         0.0   

    AverageRecharge67  High Value  
3               270.0           1  
15              343.0           1  
21              640.0           1  
24              385.0           1  
33             1932.0           1  
38              528.0           1  
48             1448.5           1  
53             1265.5           1  
54              168.0           1  
57             1147.0           1  

[10 rows x 228 columns]

### Creating Churn Customers based on Usage in September

In [133]:
ICOG9 = ((highvalue['total_ic_mou_9'] == 0) | (highvalue['total_og_mou_9'] == 0))
MB9 = ((highvalue['vol_3g_mb_9'] == 0) | (highvalue['vol_2g_mb_9'] == 0))
highvalue['churn'] = np.where((ICOG9) & (MB9),1,0)

C:\Users\cmishra\AppData\Local\Continuum\anaconda3\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  This is separate from the ipykernel package so we can avoid doing imports until


In [134]:
highvalue.head(10)

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
3      7001204172        109             0.0             0.0             0.0   
15     7002070365        109             0.0             0.0             0.0   
21     7002124215        109             0.0             0.0             0.0   
24     7001125315        109             0.0             0.0             0.0   
33     7000149764        109             0.0             0.0             0.0   
38     7000815202        109             0.0             0.0             0.0   
48     7000294396        109             0.0             0.0             0.0   
53     7002188521        109             0.0             0.0             0.0   
54     7001957064        109             0.0             0.0             0.0   
57     7001682585        109             0.0             0.0             0.0   

   last_date_of_month_6 last_date_of_month_7 last_date_of_month_8  \
3             6/30/2014            7/31/2014            8/31/2014   
15            6/30/2014            7/31/2014            8/31/2014   
21            6/30/2014            7/31/2014            8/31/2014   
24            6/30/2014            7/31/2014            8/31/2014   
33            6/30/2014            7/31/2014            8/31/2014   
38            6/30/2014            7/31/2014            8/31/2014   
48            6/30/2014            7/31/2014            8/31/2014   
53            6/30/2014            7/31/2014            8/31/2014   
54            6/30/2014            7/31/2014            8/31/2014   
57            6/30/2014            7/31/2014            8/31/2014   

   last_date_of_month_9    arpu_6  ...    fb_user_8  fb_user_9   aon  \
3             9/30/2014   221.338  ...          NaN        NaN  2491   
15            9/30/2014    31.000  ...          NaN        NaN   270   
21            9/30/2014   514.453  ...          NaN        NaN   720   
24            9/30/2014   422.050  ...          NaN        NaN  3072   
33            9/30/2014   977.020  ...          1.0        1.0  1328   
38            9/30/2014   363.987  ...          NaN        0.0  1155   
48            9/30/2014  1873.271  ...          NaN        NaN  1265   
53            9/30/2014   978.077  ...          0.0        0.0   697   
54            9/30/2014    82.866  ...          NaN        NaN   538   
57            9/30/2014  1201.298  ...          1.0        1.0  3106   

    aug_vbc_3g  jul_vbc_3g  jun_vbc_3g  sep_vbc_3g  AverageRecharge67  \
3         0.00        0.00        0.00         0.0              270.0   
15        0.00        0.00        0.00         0.0              343.0   
21        0.00        0.00        0.00         0.0              640.0   
24        0.00        0.00        0.00         0.0              385.0   
33      358.10     2706.83     1714.67         0.0             1932.0   
38        0.00        0.00        0.00         0.0              528.0   
48        0.00        0.00        0.00         0.0             1448.5   
53        0.00        0.00        0.00         0.0             1265.5   
54        0.00        0.00        0.00         0.0              168.0   
57     1799.09     3649.39     1892.32         0.0             1147.0   

    High Value  churn  
3            1      0  
15           1      0  
21           1      0  
24           1      0  
33           1      0  
38           1      0  
48           1      0  
53           1      0  
54           1      0  
57           1      0  

[10 rows x 229 columns]

#### Table for Customers who have Churned already only 295 out of 29010 entries

In [138]:
highvaluechurner = highvalue[highvalue['churn']==1]
highvaluechurner.head(10)
highvaluechurner.describe()


mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  \
count   2.950000e+02      295.0           253.0           253.0   
mean    7.001208e+09      109.0             0.0             0.0   
std     7.542685e+05        0.0             0.0             0.0   
min     7.000030e+09      109.0             0.0             0.0   
25%     7.000537e+09      109.0             0.0             0.0   
50%     7.001119e+09      109.0             0.0             0.0   
75%     7.001928e+09      109.0             0.0             0.0   
max     7.002409e+09      109.0             0.0             0.0   

       loc_ic_t2o_mou       arpu_6       arpu_7       arpu_8       arpu_9  \
count           253.0   295.000000   295.000000   295.000000   295.000000   
mean              0.0   547.711810   528.235800   552.210614   579.359234   
std               0.0   587.043434   527.800218   538.736655   433.653623   
min               0.0   -53.533000  -369.945000  -345.129000  -347.655000   
25%               0.0   180.776000   170.388500   214.160000   357.864000   
50%               0.0   375.444000   396.556000   425.310000   457.140000   
75%               0.0   729.222500   739.523500   681.398500   639.948500   
max               0.0  3872.514000  4038.527000  3911.500000  3486.400000   

       onnet_mou_6  ...     fb_user_8   fb_user_9          aon   aug_vbc_3g  \
count   229.000000  ...    108.000000  102.000000   295.000000   295.000000   
mean    243.810742  ...      0.916667    0.921569  1017.427119    29.991932   
std     481.458628  ...      0.277674    0.270177   882.352121   145.414884   
min       0.000000  ...      0.000000    0.000000   180.000000     0.000000   
25%       5.630000  ...      1.000000    1.000000   404.500000     0.000000   
50%      65.530000  ...      1.000000    1.000000   709.000000     0.000000   
75%     270.890000  ...      1.000000    1.000000  1283.500000     0.000000   
max    4480.530000  ...      1.000000    1.000000  4282.000000  1733.920000   

        jul_vbc_3g   jun_vbc_3g  sep_vbc_3g  AverageRecharge67  High Value  \
count   295.000000   295.000000  295.000000         295.000000       295.0   
mean     74.828271    98.326203    3.490034         631.979661         1.0   
std     331.452228   420.940449   44.024860         621.157787         0.0   
min       0.000000     0.000000    0.000000           0.000000         1.0   
25%       0.000000     0.000000    0.000000         240.250000         1.0   
50%       0.000000     0.000000    0.000000         460.500000         1.0   
75%       0.000000     0.000000    0.000000         874.000000         1.0   
max    3395.880000  4295.020000  727.560000        3699.000000         1.0   

       churn  
count  295.0  
mean     1.0  
std      0.0  
min      1.0  
25%      1.0  
50%      1.0  
75%      1.0  
max      1.0  

[8 rows x 217 columns]

In [139]:
highvalue9d = highvalue[highvalue.columns.drop(list(highvalue.filter(regex='_9')))]

In [140]:
highvalue9d.head(10)

mobile_number  circle_id  loc_og_t2o_mou  std_og_t2o_mou  loc_ic_t2o_mou  \
3      7001204172        109             0.0             0.0             0.0   
15     7002070365        109             0.0             0.0             0.0   
21     7002124215        109             0.0             0.0             0.0   
24     7001125315        109             0.0             0.0             0.0   
33     7000149764        109             0.0             0.0             0.0   
38     7000815202        109             0.0             0.0             0.0   
48     7000294396        109             0.0             0.0             0.0   
53     7002188521        109             0.0             0.0             0.0   
54     7001957064        109             0.0             0.0             0.0   
57     7001682585        109             0.0             0.0             0.0   

   last_date_of_month_6 last_date_of_month_7 last_date_of_month_8    arpu_6  \
3             6/30/2014            7/31/2014            8/31/2014   221.338   
15            6/30/2014            7/31/2014            8/31/2014    31.000   
21            6/30/2014            7/31/2014            8/31/2014   514.453   
24            6/30/2014            7/31/2014            8/31/2014   422.050   
33            6/30/2014            7/31/2014            8/31/2014   977.020   
38            6/30/2014            7/31/2014            8/31/2014   363.987   
48            6/30/2014            7/31/2014            8/31/2014  1873.271   
53            6/30/2014            7/31/2014            8/31/2014   978.077   
54            6/30/2014            7/31/2014            8/31/2014    82.866   
57            6/30/2014            7/31/2014            8/31/2014  1201.298   

      arpu_7  ...    fb_user_7  fb_user_8   aon  aug_vbc_3g  jul_vbc_3g  \
3    251.102  ...          NaN        NaN  2491        0.00        0.00   
15   510.465  ...          NaN        NaN   270        0.00        0.00   
21   597.753  ...          NaN        NaN   720        0.00        0.00   
24   359.730  ...          NaN        NaN  3072        0.00        0.00   
33  2362.833  ...          1.0        1.0  1328      358.10     2706.83   
38   486.558  ...          NaN        NaN  1155        0.00        0.00   
48   575.927  ...          NaN        NaN  1265        0.00        0.00   
53  1141.296  ...          0.0        0.0   697        0.00        0.00   
54   185.157  ...          NaN        NaN   538        0.00        0.00   
57   687.704  ...          1.0        1.0  3106     1799.09     3649.39   

    jun_vbc_3g  sep_vbc_3g  AverageRecharge67  High Value  churn  
3         0.00         0.0              270.0           1      0  
15        0.00         0.0              343.0           1      0  
21        0.00         0.0              640.0           1      0  
24        0.00         0.0              385.0           1      0  
33     1714.67         0.0             1932.0           1      0  
38        0.00         0.0              528.0           1      0  
48        0.00         0.0             1448.5           1      0  
53        0.00         0.0             1265.5           1      0  
54        0.00         0.0              168.0           1      0  
57     1892.32         0.0             1147.0           1      0  

[10 rows x 175 columns]